src/Product/FilteredProductList.php

```diff
@@ -0,0 +1,90 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\Product;
+
+use Countable;
+use WC_Product;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class FilteredProductList
+ *
+ * A list of filtered products and their total count before filtering.
+ *
+ * @since x.x.x
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\Product
+ */
+class FilteredProductList implements Countable {
```

Comment:
> I considered that but didn't want to add too many functions to the class which we don't even use. It will be less relevant if we add a ProductFilter class since there will be less functions handling a filtered product list.

We could also use PHP's own `ArrayIterator` here that implements all those interfaces (including `Countable`) and the necessary methods. 

The `FilteredProductList` could implement the `IteratorAggregate` interface and then return an instance of `ArrayIterator`. This way we get all the benefits without overriding those functions ourselves. 

We could get rid of the extra `->get` call and just include the custom methods like `get_unfiltered_count` and `get_product_ids`.

Reply:
I'd like to get rid of the extra get calls, but regardless of which class we extend or implement we will always get an object back. When we get this object back we need to pass it's list on to functions like these:

```
AbstractBatchedActionSchedulerJob::schedule_process_action( array $items )
AbstractBatchedActionSchedulerJob::get_batch( int $batch_number ): array
ProductSyncer::update( array $products ): BatchProductResponse
```

Do you think we should remove the type restriction for those functions? Or should we have another way to pass it an array instead of an object?

Reply:
Yeah, you're right. We need those type checks and it's not worth removing them just to satisfy this particular case.

We could modify the types to be a general class like `ArrayCollection` and have the `FilteredProductList` extend that but it's probably best not to make things more complicated than we need to.

Even if we don't remove the `get` calls, it might be a good idea to make this iterable because it is in fact designed for holding an array. We can cast it to `array` if we don't want to deal with type hints. Given the effects you mentioned, it's not very critical to make this change. Totally up to you.

Reply:
Thanks for the clarifications. My gut feeling tells me to leave it as is. While making it iterable sounds useful, we don't actually iterate through the list, we just pass it on to another function as an array before iterating.
Casting to an array is kind of the same as using a `->get` function because PHP has discussed a magic method for casting to array but it's never been implemented, so we have to rely on a custom function to get the array.